awsx/ec2/subnetDistributorNew.ts

```diff
@@ -16,17 +16,52 @@
 // and used in accordance with MPL v2.0 license
 
 import * as pulumi from "@pulumi/pulumi";
-import { SubnetSpecInputs, SubnetTypeInputs } from "../schema-types";
+import { SubnetSpecInputs } from "../schema-types";
 import { Netmask } from "netmask";
+import { SubnetSpec, SubnetSpecPartial } from "./subnetSpecs";
 
-export interface SubnetSpec {
-  cidrBlock: string;
-  type: SubnetTypeInputs;
-  azName: string;
-  subnetName: string;
-  tags?: pulumi.Input<{
-    [key: string]: pulumi.Input<string>;
-  }>;
+// Like getSubnetSpecs but tolerates partially known vpcCidr.
+export function getSubnetSpecsWithPartialCidr(
+  vpcName: string,
+  vpcCidr: pulumi.Input<string>,
+  azNames: string[],
+  subnetInputs: SubnetSpecInputs[] | undefined,
+  azCidrMask?: number,
+): SubnetSpecPartial[] {
+  const allocatedCidrBlocks = allocateSubnetCidrBlocksInput(
+    vpcName,
+    vpcCidr,
+    azNames,
+    subnetInputs,
+    azCidrMask,
+  );
+  const subnetSpecs = subnetInputs ?? defaultSubnetInputsBare();
+  return azNames.flatMap((azName, azIndex) => {
```

Comment:
Can we combine this with the `allocateSubnetCidrBlocksInput`/`allocationSubnetCidrBlocks` function? It looks like we perform the same type of loop a bunch of different places which makes it a little hard to follow what is happening.

Reply:
Yeah, I don't know how to make this cleaner. allocateSubnetCidrBlocksInput is very intentionally broken out. Our problem here is that we're trying to keep the allocateSubnetCidrBlocks preserve the original algo as-is. But we need to interface with this algo when`vpcCidr: pulumi.Input<string>` is not known yet. Even in this case where it's not know yet, we can't assume that the entire set of `SubnetSpecPartial[]` is unknown - we know how many there will be, and we know parts of information about them right away at preview, just not their respective cidrBlock values.

This indirection is accomplished by the means of `Record<SubnetAllocationID, { cidrBlock: pulumi.Input<string> }>` helper data structure.

- allocateSubnetCidrBlocks is needed to preserve original algo in as simple code as possible
- allocateSubnetCidrBlocksInput is needed to do tricks to make allocateSubnetCidrBlocks callable with vpcCidr that is now an Input but return a Record, not an `Output<Record>` - tricks are used to not lose information
- getSubnetSpecsWithPartialCidr/getSubnetSpecs is needed to do the rest of SubnetSpecPartial computation, not just the cidrBlock

Reply:
What if we used union types and type checking? Something like this:

```typescript
export function getSubnetSpecsWithPartialCidr(
  vpcName: string,
  vpcCidr: pulumi.Input<string>,
  azNames: string[],
  subnetInputs: SubnetSpecInputs[] | undefined,
  azCidrMask?: number,
): SubnetSpecPartial[] {
  let alloc:
    | pulumi.Output<Record<SubnetAllocationID, { cidrBlock: pulumi.Input<string> }>>
    | Record<SubnetAllocationID, { cidrBlock: pulumi.Input<string> }>;
  if (typeof vpcCidr === "string") {
    alloc = allocateSubnetCidrBlocks(vpcName, vpcCidr, azNames, subnetInputs, azCidrMask);
  } else {
    alloc = pulumi
      .output(vpcCidr)
      .apply((vpcCidr) =>
        allocateSubnetCidrBlocks(vpcName, vpcCidr, azNames, subnetInputs, azCidrMask),
      );
  }
  const subnetSpecs = subnetInputs ?? defaultSubnetInputsBare();
  return azNames.flatMap((azName, azIndex) => {
    const azNum = azIndex + 1;
    return subnetSpecs.map((subnetSpec, subnetIndex) => {
      const subnetAllocID = subnetAllocationID(vpcName, subnetSpec, azNum, subnetIndex);
      let cidrBlock: pulumi.Input<string>;
      if (pulumi.Output.isInstance(alloc)) {
        const allocOutput = alloc as pulumi.Output<
          Record<SubnetAllocationID, { cidrBlock: pulumi.Input<string> }>
        >;
        cidrBlock = allocOutput.apply((a) => pulumi.output(a[subnetAllocID].cidrBlock));
      } else {
        cidrBlock = alloc[subnetAllocID].cidrBlock as pulumi.Input<string>;
      }
      return {
        cidrBlock: cidrBlock,
        type: subnetSpec.type,
        azName,
        subnetName: subnetName(vpcName, subnetSpec, azNum),
        tags: subnetSpec.tags,
      };
    });
  });
}
```

Reply:
This is a nice idea, I think you are right. This lets me delete one of the functions. I opted for an even simpler form as I think we do not need to `if (typeof vpcCidr === "string") {` for cidrBlock we can just completely commit to pulumi.Output without loss of pertinent information in this case. Have a look at the latest version and thanks for pushing on this 🙏